src/main/java/io/fabric8/maven/docker/service/QueryService.java

```diff
@@ -181,30 +181,42 @@ public boolean hasImage(String name) throws DockerAccessException {
      * @throws DockerAccessException
      * @throws MojoExecutionException
      */
-    public boolean imageRequiresAutoPull(String mode, String imageName, boolean always)
+    public boolean imageRequiresAutoPull(String mode, String imageName, boolean always, Set<String> previouslyPulled)
         throws DockerAccessException, MojoExecutionException {
 
         // The logic here is like this (see also #96):
-        // If the image is not available and mode is either ON or ALWAYS --> pull
+        // If the image is not available and mode is one of: ON, ALWAYS, ONCE --> pull
         // If mode == ALWAYS and no build config is available (so its a pulled-image anyway) --> pull
         // otherwise: don't pull
         AutoPullMode autoPullMode = AutoPullMode.fromString(mode);
-        if (pullIfNotPresent(autoPullMode, imageName) ||
-            alwaysPull(autoPullMode, always)) {
+        if (imageRequiresPull(autoPullMode, imageName, always, previouslyPulled)) {
             return true;
-        } else {
-            if (!hasImage(imageName)) {
-                throw new MojoExecutionException(
-                        String.format("No image '%s' found, Please enable 'autoPull' or pull image '%s' yourself (docker pull %s)",
-                                imageName, imageName, imageName));
-            }
+        }
+
+        if (hasImage(imageName)) {
             return false;
         }
+
+        throw new MojoExecutionException(
+                String.format("No image '%s' found, Please enable 'autoPull' or pull image '%s' yourself (docker pull %s)",
+                        imageName, imageName, imageName));
+    }
+
+    private boolean imageRequiresPull(AutoPullMode autoPullMode, String imageName, boolean always, Set<String> previouslyPulled)
+            throws DockerAccessException {
+
+        if (autoPullMode == AutoPullMode.ONCE && previouslyPulled.contains(imageName)) {
+            return false;
+        }
+
+        previouslyPulled.add(imageName);
```

Comment:
I would update the set here as a side effect. Instead its better to do this when the auto pull actually happens.